Middleware support

bonahona opened the pull request:
Added possibility to install interceptors into the workflow.

bonahona committed:
Added some sort of worker interceptor that can execute arbitrary code while encapsulating the worker function. WIP.

bonahona committed:
Lots of changes to how the featues, now called interceptors, works.
Each interceptor can now (and most now) execute the next step deliberatly.

bonahona committed:
Added more test cases for interceptors.

bonahona committed:
Added possibility to override the result of a JobState via an interceptor.

bonahona committed:
Removed lots of debug println lines in the test cases.

bonahona committed:
Removed TODOs as that code works.

williamhogman commented on src/main/kotlin/WorkerBuilder.kt:
```diff
@@ -1,10 +1,9 @@
 package franz
 
-import franz.engine.ConsumerActor
+import com.sun.xml.internal.ws.api.pipe.FiberContextSwitchInterceptor
```
Remove plz

williamhogman commented on src/main/kotlin/JobState.kt:
```diff
@@ -78,21 +78,39 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?) {
-            this.status = newStatus
-            msg?.let { log.debug("Failed on: $it") }
+    fun process(newStatus: JobStatus, predicate: (U) -> Boolean, msg: String? = null): JobState<U> {
+        val lastInterceptor = WorkerInterceptor {
```
Could we avoid this allocation in some easy way? If no then fine

bonahona replied:
Not that I know of. This was the easiet way to put the execution of the desired worker function last.

williamhogman reviewed (approved):
Only minor things

williamhogman commented:
Great tests @bonahona

williamhogman commented:
Lets add @mantono too for review

bonahona committed:
Removed unneccesary import.
Removed old comment.

mantono reviewed (changes requested):
You cannot remove modifier `inline` on functions 
`require`, `execute`, `advanceIf`, `sideEffect`, `end`, `map` and `process`, since  now we cannot pass suspend functions to them, which is not acceptable.

williamhogman commented:
>You cannot remove modifier inline on functions
require, execute, advanceIf, sideEffect, end, map and process, since now we cannot pass suspend functions to them, which is not acceptable.

Perhaps `crossinline` is required  at some places

mantono commented:
```
    private suspend fun suspendingFunction() { "Lets do nothing" }

    @Test
    fun testMapWithSuspendFunctionWillCompile() {
        runBlocking {
            jobOne.map { suspendingFunction() }
        }
    }
```

This test case (if added) would not compile now

mantono commented:
@williamhogman Yes possibly, I am not sure what prompted the inline modifier to be removed in the first place, but I assume it was required in order for the interceptors to work?

bonahona committed:
Readded inline keywords to jobstate.
Added keywords crossinline to workerfunction predicates.

mantono committed:
Added test cases that checks that Franz functions can handle suspend functions.

mantono commented:
@bonahona @williamhogman `crossinline` did not work :( Added "test case" to check against.

bonahona committed:
Added inline to additional functions in JobState.kt.
Added testcase just to make sure the build fails if suspend functions are not possible to use in map/execure/require functions.

bonahona committed:
Merge branch 'middleware-support' of github.com:zensum/franz into middleware-support

bonahona committed:
WIP on suspend functions.

bonahona committed:
Made all of JobState work as suspend functions.
Added test cases for some more worker.

bonahona committed:
Added new test case for all cases of piped actions job state.

bonahona committed:
Changed some more functions in JobState for inlined to suspend.

mantono commented on src/main/kotlin/JobState.kt:
```diff
@@ -51,7 +51,7 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?, val int
      * If the [JobState] is non-terminal mark it as a success
      */
-    fun success(): JobState<U> = also { it.status = JobStatus.Success }
+    suspend fun success(): JobState<U> = also { it.status = JobStatus.Success }
```
This one does not need the suspend modifier as it can never execute a suspending function.

mantono commented on src/main/kotlin/JobState.kt:
```diff
@@ -60,7 +60,7 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?, val int
      * prior to this function call).
      * */
-    inline fun end(predicate: (U) -> Boolean): JobStatus {
+    fun end(predicate: (U) -> Boolean): JobStatus {
```
This one does need suspend however (you probably aware of that and just missed it) ;)

mantono commented on src/main/kotlin/JobState.kt:
```diff
@@ -72,7 +72,7 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?, val int
 
     @JvmName("endNullary")
-    fun end(): JobStatus {
+    suspend fun end(): JobStatus {
```
suspend modifier is not needed here either.

bonahona committed:
Added/removed suspend modifier on some methods in JobState where needed.

mantono commented on src/main/kotlin/JobState.kt:
```diff
@@ -60,7 +60,7 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?) {
      * prior to this function call).
      * */
-    inline fun end(predicate: (U) -> Boolean): JobStatus {
+    suspend fun end(predicate: (U) -> Boolean): JobStatus {
```
We need a `suspend` on `predicate`.

mantono commented on src/main/kotlin/JobState.kt:
```diff
@@ -72,27 +72,44 @@ class JobState<U: Any> @PublishedApi internal constructor(val value: U?) {
     }
 
-    inline fun <R: Any> map(transform: (U) -> R): JobState<R> {
+    suspend fun <R: Any> map(transform: (U) -> R): JobState<R> {
```
This function need to be changed to
```
    suspend fun <R: Any> map(transform: suspend (U) -> R): JobState<R> {
        val transFormedVal: R? = when (inProgress()) {
            true -> value?.let { transform(it) }
            false -> null
        }
```

mantono reviewed (changes requested):
I use the following test cases to ensure that we can pass both suspend and non suspending functions to the function;
```
    private suspend fun suspendingFunction(): Boolean = true
    private fun nonSuspendingFunction(): Boolean = true

    @Test
    fun testMapWithSuspendFunctionWillCompile() {
        runBlocking {
            jobOne
                .advanceIf { suspendingFunction() }
                .execute { suspendingFunction() }
                .require { suspendingFunction() }
                .sideEffect { suspendingFunction() }
                .map { suspendingFunction() }
                .end { suspendingFunction() }
        }
    }

    @Test
    fun testMapWithNonSuspendFunctionWillCompile() {
        runBlocking {
            jobOne
                .advanceIf { nonSuspendingFunction() }
                .execute { nonSuspendingFunction() }
                .require { nonSuspendingFunction() }
                .sideEffect { nonSuspendingFunction() }
                .map { nonSuspendingFunction() }
                .end { nonSuspendingFunction() }
        }
    }
```

It they both pass, then I think we are all good.

bonahona committed:
Reverted map method back to an inline method as it does use interceptors.

mantono committed:
Make function end that takes a predicate lambda have suspend modifier so it compiles.
